lib/dao.js

```diff
@@ -2407,6 +2577,201 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = this.getConnector();
+  assert(typeof connector.replaceById === 'function',
+          'replaceById() must be implemented by the connector');
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  // Make sure id(s) cannot be changed
+  var idNames = Model.definition.idNames();
+  for (var i = 0, n = idNames.length; i < n; i++) {
+    var IdName = idNames[i];
+    if (data[IdName] !== undefined && !idEquals(data[IdName], inst[IdName])) {
+      var err = new Error('id property (' + IdName + ') ' +
+              'cannot be updated from ' + inst[IdName] + ' to ' + data[IdName]);
+      err.statusCode = 400;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+  }
+  var context = {
+    Model: Model,
+    instance: inst,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    if (strict) {
+      applyStrictCheck(Model, strict, data, inst, validateAndCallConnector);
+    } else {
+      validateAndCallConnector(null, data);
+    }
+
+    function validateAndCallConnector(err, data) {
+      if (err) return cb(err);
+      data = removeUndefined(data);
+      // update instance's properties    
+      inst.setAttributes(data);
+
+      var doValidate = true;
+      if (options.validate === undefined) {
+        if (Model.settings.automaticValidation !== undefined) {
+          doValidate = Model.settings.automaticValidation;
+        }
+      } else {
+        doValidate = options.validate;
+      }
+
+      if (doValidate){
+        inst.isValid(function (valid) {
+          if (!valid) {
+            return cb(new ValidationError(inst), inst);
+          }
+          callConnector();
+        }, data);
+      } else {
+          callConnector();
+      }
+
+      function callConnector() {
+        var idNames = Model.definition.idNames();
+        // Convert idNames to an object since when we check ids
+        // the complexity would be O(1) in Object as opposed to
+        // an array which would be O(N)
+        // Please feel free to let me know if you dont like the idea
+        idNames = idNames.reduce(function(result, currentItem) {
+          result[currentItem] = currentItem;
+          return result;
+        }, {});
+
+        var propKeys = Object.keys(Model.definition.properties);
+        for (var i = 0; i < propKeys.length; i++) {
+          var p = propKeys[i];
+          if (idNames[p])
+            continue;
+          inst[p] = null;
```

Comment:
Are the connectors prepared for the fact that they will receive data without `id` fields filled in, i.e. they have to add them explicitely?